File: modules/auth-engine/src/main/java/com/kamcci/modules/auth/engine/support/UserDetailArgumentResolver.java
package com.kamcci.modules.auth.engine.support;

import com.kamcci.modules.auth.control.annotation.UserEmail;
import com.kamcci.modules.auth.control.annotation.UserId;
import com.kamcci.modules.auth.control.annotation.UserRole;
import com.kamcci.modules.auth.control.enumeration.UserRoleType;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.core.MethodParameter;
import org.springframework.lang.Nullable;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Component
@Qualifier("userDetail")
public class UserDetailArgumentResolver implements HandlerMethodArgumentResolver {
    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.getParameterAnnotation(UserId.class) != null || parameter.getParameterAnnotation(UserEmail.class) != null || parameter.getParameterAnnotation(UserRole.class) != null;
    }

    @Nullable
    @Override
    public Object resolveArgument(MethodParameter parameter, @Nullable ModelAndViewContainer mavContainer,
                                  NativeWebRequest webRequest, @Nullable WebDataBinderFactory binderFactory) {

        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        boolean hasUserIdAnnot = parameter.getParameterAnnotation(UserId.class) != null;
        boolean hasUserEmailAnnot = parameter.getParameterAnnotation(UserEmail.class) != null;
        boolean isAnonymousUser = authentication.getPrincipal().equals("anonymousUser");

        if(hasUserIdAnnot) {
            // @UserId 처리
            Map<String, Object> details = (Map<String, Object>) authentication.getDetails();
            return isAnonymousUser ? null : UUID.fromString(details.get(UserId.ATTR_NAME).toString());
        } else if(hasUserEmailAnnot) {
            // @UserEmail 처리
            return isAnonymousUser ? "" : authentication.getPrincipal().toString();
        } else {
            // @UserRole 처리
            if(isAnonymousUser) {
                return new ArrayList<UserRoleType>();
            } else {
                List<String> roles = new ArrayList<>();
                authentication.getAuthorities().forEach(authority -> roles.add(authority.getAuthority()));

                List<UserRoleType> roleTypeList = new ArrayList<>();
                for(String role : roles) {
                    UserRoleType roleType = UserRoleType.valueOf(role);
                    roleTypeList.add(roleType);
                }
                return roleTypeList;
            }
        }
    }

}

File: modules/auth-engine/src/main/java/com/kamcci/modules/auth/engine/support/UserDetailArgumentResolver.kt
//package com.kamcci.modules.auth.engine.support
//
//import com.kamcci.modules.auth.control.annotation.UserEmail
//import com.kamcci.modules.auth.control.annotation.UserId
//import com.kamcci.modules.auth.control.annotation.UserRole
//import com.kamcci.modules.auth.control.enumeration.UserRoleType
//import org.springframework.beans.factory.annotation.Qualifier
//import org.springframework.core.MethodParameter
//import org.springframework.security.core.context.SecurityContextHolder
//import org.springframework.stereotype.Component
//import org.springframework.web.bind.support.WebDataBinderFactory
//import org.springframework.web.context.request.NativeWebRequest
//import org.springframework.web.method.support.HandlerMethodArgumentResolver
//import org.springframework.web.method.support.ModelAndViewContainer
//import java.util.*
//
//@Component
//@Qualifier("userDetail")
//class UserDetailArgumentResolver : HandlerMethodArgumentResolver {
//    override fun supportsParameter(parameter: MethodParameter) =
//        parameter.getParameterAnnotation(UserId::class.java) != null ||
//                parameter.getParameterAnnotation(UserEmail::class.java) != null ||
//                parameter.getParameterAnnotation(UserRole::class.java) != null
//
//    override fun resolveArgument(
//        parameter: MethodParameter,
//        mavContainer: ModelAndViewContainer?,
//        webRequest: NativeWebRequest,
//        binderFactory: WebDataBinderFactory?
//    ): Any {
//        val authentication = SecurityContextHolder.getContext().authentication
//
//        val hasUserIDAnnot = parameter.getParameterAnnotation(UserId::class.java) != null
//        val hasUserEmailAnnot = parameter.getParameterAnnotation(UserEmail::class.java) != null
//        val isAnonymousUser = authentication.principal == "anonymousUser"
//
//        return when {
//            // @UserID 적용 및 인증된 사용자
//            hasUserIDAnnot -> {
//                if (isAnonymousUser) 0
//                else authentication.details as UUID
//            }
//
//            // @UserEmail 적용 및 인증된 사용자
//            hasUserEmailAnnot -> {
//                if (isAnonymousUser) ""
//                else authentication.principal.toString()
//            }
//
//            // @UserRole 적용 및 인증된 사용자
//            else -> {
//                if (isAnonymousUser) listOf<UserRoleType>()
//                else {
//                    val roles = authentication.authorities.map { it.authority }
//                    val roleTypeList: MutableList<UserRoleType?> = mutableListOf()
//                    roles.forEach { role ->
//                        roleTypeList.add(UserRoleType.entries.find { it.name == role })
//                    }
//                    roleTypeList
//                }
//            }
//        }
//    }
//}